### PR TITLE
[0.3.4]: feat: move `background_round` outside of `abstract_round_abci`

### PR DESCRIPTION
**Pull request by 0xArdi**
## Proposed changes

This PR moves the `background_round` (termination_round) outside the `asbtract_round_abci`.

## Types of changes

What types of changes does your code introduce? (A **breaking change** is a fix or feature that would cause existing functionality and APIs to not work as expected.)
_Put an `x` in the box that applies_

- [ ] Non-breaking fix (non-breaking change which fixes an issue)
- [ ] Breaking fix (breaking change which fixes an issue)
- [x] Non-breaking feature (non-breaking change which adds functionality)
- [ ] Breaking feature (breaking change which adds functionality)
- [ ] Refactor (non-breaking change which changes implementation)
- [ ] Messy (mixture of the above - requires explanation!)

## Checklist

_Put an `x` in the boxes that apply._

- [x] I have read the [CONTRIBUTING](../blob/main/CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `main` branch (left side). Also you should start your branch off our `main`.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have locally run services that could be impacted and they do not present failures derived from my changes

### CODE REVIEW

File: packages/valory/skills/abstract_round_abci/base.py [hunk 1]
```diff
@@ -1835,7 +1830,9 @@ class AbciApp(
     final_states: Set[AppState] = set()
     event_to_timeout: EventToTimeout = {}
     cross_period_persisted_keys: List[str] = []
-
+    termination_round_cls: Optional[AppState] = None
+    termination_transition_function: Optional[AbciAppTransitionFunction] = None
+    termination_event: Optional[EventType] = None
```
Comment (0xArdi) [inline packages/valory/skills/abstract_round_abci/base.py:1835]: we could have a termination class to handle this

**Review by Adamantios — COMMENTED** [review]
Looks good so far, this is shaping up nicely.

File: packages/valory/skills/abstract_round_abci/behaviours.py [hunk 2]
```diff
@@ -156,6 +153,7 @@ class AbstractRoundBehaviour(
     abci_app_cls: Type[AbciApp[EventType]]
     behaviours: AbstractSet[BehaviourType]
     initial_behaviour_cls: BehaviourType
+    termination_behaviour_cls: Optional[BehaviourType]
```
[the remark this answers is not in the document]
Comment (Adamantios) [reply packages/valory/skills/abstract_round_abci/behaviours.py:156]: IMO it would make sense, for now, to keep this as `termination`. 

If we extend it in the future and support multiple background behaviours then we could rename the attribute.

File: packages/valory/skills/abstract_round_abci/base.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (0xArdi) [reply packages/valory/skills/abstract_round_abci/base.py:1835]: somethink like:
```
AnyAbciApp = with_termination(
    chain(
        (
            RegistrationAbci.AgentRegistrationAbciApp,
            SafeDeploymentAbci.SafeDeploymentAbciApp,
            TransactionSubmissionAbci.TransactionSubmissionAbciApp,
            ResetAndPauseAbci.ResetPauseABCIApp,
        ),
        abci_app_transition_mapping,
    )
)
```

[the remark this answers is not in the document]
Comment (DavidMinarsch) [reply packages/valory/skills/abstract_round_abci/base.py:1835]: I like the idea of applying it at the composition level. That's what I was hoping you'd do. I suggest a slightly different syntax:

```
ChainedAbciApp = chain().add_termination(abci_app: Type[AbciApp] = TerminationAbciApp, transition_function: Type[AbciAppTransitionFunction] = ..)
```
where `ChainedTerminationAbciApp` can be e.g. the TermiationAbci + TransactionSettlementAbci chained.

Then this can be fined for instance here: https://github.com/valory-xyz/price-oracle/blob/main/packages/valory/skills/oracle_abci/composition.py#L73-L83

File: packages/valory/skills/abstract_round_abci/base.py [hunk 3]
```diff
@@ -2547,16 +2535,19 @@ def _update_round(self) -> None:
...
+            ] = None
+        if self.abci_app.is_termination_set:
+            termination_result: Optional[
+                Tuple[BaseSynchronizedData, Any]
+            ] = self.abci_app.termination_round.end_block()
+            if termination_result is not None and not self._termination_called:
+                # when the termination round returns, it takes priority over normal rounds
+                # because the TerminationRound never ends, we should only take into account
+                # its response only once
+                self._termination_called = True
```
[the remark this answers is not in the document]
Comment (DavidMinarsch) [reply packages/valory/skills/abstract_round_abci/base.py:2549]: When would that be needed? It's a one-way operation as currently defined.

File: packages/valory/skills/abstract_round_abci/base.py [hunk 4]
```diff
@@ -2106,14 +2097,11 @@ def process_event(
...
-                raise ValueError(
-                    "Termination majority was reached, "
-                    "but the transition to `transaction_settlement_abci` failed."
-                )
-
+        if self._is_termination_set and event == self.termination_event:
+            next_round_cls = self.termination_transition_function[self.termination_round_cls].get(
+                event, None
+            )
+            self.transition_function = self.termination_transition_function
```
Comment (DavidMinarsch) [inline packages/valory/skills/abstract_round_abci/base.py:2104]: Add a comment in code here that entering the termination is a one-way operation.

File: packages/valory/skills/abstract_round_abci/behaviours.py [hunk 5]
```diff
@@ -168,7 +166,8 @@ def __init__(self, **kwargs: Any) -> None:
         ] = self._get_round_to_behaviour_mapping(self.behaviours)
 
         self.current_behaviour: Optional[BaseBehaviour] = None
-        self.__termination_behaviour: Optional[BaseBehaviour] = None
+        self.termination_behaviour: Optional[BaseBehaviour] = None
```
Comment (DavidMinarsch) [inline packages/valory/skills/abstract_round_abci/behaviours.py:169]: For clarity, we could call this `concurrent_background_behaviour`. As really it is concurrent & in the background. The termination is only in so far correct from perspective of `abstract_round_abci` that it is a one-way transition (for now). Although in the future, one could imagine that the `concurrent_background_behaviour` could itself be replaced here: https://github.com/valory-xyz/open-autonomy/pull/1471/files#diff-e0565ddcb0e5eea72afcbdb00881efedc60c4664d72af704f5d428578c68bffbR2104

File: packages/valory/skills/abstract_round_abci/behaviours.py [hunk 6]
```diff
@@ -168,7 +166,8 @@ def __init__(self, **kwargs: Any) -> None:
...
+        self._is_termination_set = self.termination_behaviour_cls is not None
```
Comment (DavidMinarsch) [inline packages/valory/skills/abstract_round_abci/behaviours.py:170]: Rather than `termination` I suggest something with `background` or `exit`. I would reserve `termination` for the actual behaviour we are implementing below. Then it is less confusing.

File: packages/valory/skills/termination_abci/behaviours.py [hunk 7]
```diff
@@ -43,7 +43,7 @@
     AsyncBehaviour,
     BaseBehaviour,
 )
-
+from packages.valory.skills.transaction_settlement_abci.payload_tools import hash_payload_to_hex
```
Comment (DavidMinarsch) [inline packages/valory/skills/termination_abci/behaviours.py:46]: At some point this should be split into rounds and behaviours. But I'd do it in a separate PR for easier diffs.

Also, you're missing the definition of the AbciApp

**Review by DavidMinarsch — COMMENTED** [review]
Nice progress!

File: packages/valory/skills/abstract_round_abci/behaviours.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (DavidMinarsch) [reply packages/valory/skills/abstract_round_abci/behaviours.py:156]: I argue for background below inside `abstract_round_abci` and for `termination` for the skill.

**Review by DavidMinarsch — APPROVED** [review]
